File: examples/diagnostics/inspect/codelab/rust/part_1/src/main.rs
// Copyright 2020 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

use {
    crate::reverser::ReverserServerFactory,
    anyhow::{Context, Error},
    fidl_fuchsia_examples_inspect::FizzBuzzMarker,
    fuchsia_async as fasync,
    fuchsia_component::{client, server::ServiceFs},
    fuchsia_syslog::{self as syslog, macros::*},
    futures::{future::try_join, FutureExt, StreamExt},
    // CODELAB: Use inspect.
};

mod reverser;

#[fasync::run_singlethreaded]
async fn main() -> Result<(), Error> {
    // [START init_logger]
    syslog::init_with_tags(&["inspect_rust_codelab", "part1"])?;
    // [END init_logger]

    // [START servicefs_init]
    let mut fs = ServiceFs::new();
    // [END servicefs_init]

    fx_log_info!("starting up...");

    // CODELAB: Initialize Inspect here.

    // Create a new Reverser Server factory.
    let reverser_factory = ReverserServerFactory::new();

    // Serve the reverser service
    // [START serve_service]
    fs.dir("svc").add_fidl_service(move |stream| reverser_factory.spawn_new(stream));
    fs.take_and_serve_directory_handle()?;
    // [END serve_service]

    // Send a request to the FizzBuzz service and print the response when it arrives.
    // [START fizzbuzz_connect]
    let fizzbuzz_fut = async move {
        let fizzbuzz = client::connect_to_service::<FizzBuzzMarker>()
            .context("failed to connect to fizzbuzz")?;
        match fizzbuzz.execute(30u32).await {
            Ok(result) => fx_log_info!("Got FizzBuzz: {}", result),
            Err(_) => {}
        };
        Ok(())
    };
    // [END fizzbuzz_connect]

    // [START servicefs_collect]
    let running_service_fs = fs.collect::<()>().map(Ok);
    // [END servicefs_collect]
    try_join(running_service_fs, fizzbuzz_fut).await.map(|((), ())| ())
}

File: examples/diagnostics/inspect/codelab/dart/part_1/lib/main.dart
// Copyright 2020 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

// TODO(https://fxbug.dev/84961): Fix null safety and remove this language version.
// @dart=2.9

import 'package:fidl_fuchsia_examples_inspect/fidl_async.dart' as fidl_codelab;
import 'package:fuchsia_logger/logger.dart';
import 'package:fuchsia_services/services.dart';
import 'package:inspect_dart_codelab_part_1_lib/reverser.dart';
// CODELAB: use inspect.

void main(List<String> args) async {
  // [START init_logger]
  setupLogger(name: 'inspect_dart_codelab', globalTags: ['part_1']);
  // [END init_logger]

  log.info('Starting up...');

  // CODELAB: Initialize Inspect here.

  // [START serve_service]
  final context = ComponentContext.create();
  context.outgoing
    ..addPublicService<fidl_codelab.Reverser>(
      ReverserImpl.getDefaultBinder(),
      fidl_codelab.Reverser.$serviceName,
    )
    ..serveFromStartupInfo();
  // [END serve_service]

  try {
    // [START connect_fizzbuzz]
    final fizzBuzz = fidl_codelab.FizzBuzzProxy();
    context.svc.connectToService(fizzBuzz);
    final result = await fizzBuzz.execute(30);
    // [END connect_fizzbuzz]
    log.info('Got FizzBuzz: $result');
  } on Exception catch (_) {}
}
